internal/controller/promotions/promotions.go

```diff
@@ -90,29 +116,12 @@ func newReconciler(
 	return r
 }
 
-func newPromotionsQueue() runtime.PriorityQueue {
-	// We can safely ignore errors here because the only error that can happen
-	// involves initializing the queue with a nil priority function, which we
-	// know we aren't doing.
-	pq, _ := runtime.NewPriorityQueue(func(left, right client.Object) bool {
-		return left.GetCreationTimestamp().Time.
-			Before(right.GetCreationTimestamp().Time)
-	})
-	return pq
-}
-
 // Reconcile is part of the main Kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
 func (r *reconciler) Reconcile(
 	ctx context.Context,
 	req ctrl.Request,
 ) (ctrl.Result, error) {
-	// We count all of Reconcile() as a critical section of code to ensure we
-	// don't start reconciling a second Promotion before lazy initialization
-	// completes upon reconciliation of the FIRST promotion.
-	r.promoQueuesByStageMu.Lock()
-	defer r.promoQueuesByStageMu.Unlock()
```

Comment:
TIL that multiple calls to `sync.Once` don't return until the one that's actually doing something returns. i.e. Built-in mutex! Nice!

Reply:
I actually didn't know about `sync.Once` until this refactor and so I think I was the one doing the learning 😉 .